### PR TITLE
Patch Glyphicons and Add Font Awesome Icons

### PR DESCRIPTION
**Pull request by Chng-Zhi-Xuan**
**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [X] Documentation update
• [X] Enhancement to an existing feature
• [X] Other, migration

Part of #298 and MarkBind/vue-strap#53

**What is the rationale for this request?**
`Glyphicons` has been deprecated from Bootstrap 4.x assets, which caused missing icons within MarkBind's components.

**What changes did you make? (Give an overview)**
- Import existing `Glyphicons` CSS file from [this repository](https://github.com/ohpyupi/glyphicons-only-bootstrap).
- Added `Font Awesome`'s set of free icons into our assets
- Reference both icons' style sheet in `page.ejs`
- Add `Font Awesome` to `userDefinedVariables` map, same as `Glyphicons`
- Update user guide in [Font Awesome](https://deploy-preview-301--markbind.netlify.com/userguide/contentauthoring#font-awesome-icons)

Discussion |
--- |
Font Awesome was added as a (better) replacement to Glyphicons, as there is no license that explicitly mentions the use of Glyphicons in Bootstrap 4.x (only mentioned dropping it). Further investigation is required, and if found that it violates Glyphicon's terms of use, we will remove it from our application. | 
**Part Two** After team discussion, we decided that importing Glyphicon's CSS file from Bootstrap 3.x is within fair use. Thus we will continue supporting Glyphicons in MarkBind. Acknowledgements of any other projects or assets we use will be done in our repository's README page.

**Is there anything you'd like reviewers to focus on?**
- New `getIconsMap` method, to collate all icon assets.
- Phrasing of user guide 

**Testing instructions:**
- Ensure you have Bootstrap's 4.x `.min.js` file in your asset folder
- Check that Glyphicons and Font Awesome icons work by using `{{glyphicon_home}}` and `{{fab_playstation}}` in any `index.md` before serving your site.

### CODE REVIEW

**Comment by damithc** [comment]
In the documentation, we can add something like this:
> ❗️ The glyphicons which MarkBind supports by default are the ones available for free from http://glyphicons.com/ under the are released under the [Creative Commons Attribution 3.0 Unported (CC BY 3.0)](). By using those glyphicons in your MarkBind website, you agree to their [license terms](http://glyphicons.com/license/). 

We can also add something like this in our MarkBind website footer (to further comply with glyphicon's license terms).
> Acknowledgements: [VueStrap](), [Glyphicons](), [FontAwesome]()

Apparently, there is no explicit clause against packaging free icons in another tool. That clause is in the non-free versions only.

**Comment by Chng-Zhi-Xuan** [comment]
**Update**
- Update PR description.
- Ready for review.

File: lib/Site.js [hunk 1]
```diff
@@ -32,6 +32,7 @@ const TEMPLATE_SITE_ASSET_FOLDER_NAME = 'markbind';
 
 const FAVICON_DEFAULT_PATH = 'favicon.ico';
 const FOOTER_PATH = '_markbind/footers/footer.md';
+const FONT_AWESOME_PATH = 'asset/font-awesome.csv';
```
Comment (yamgent) [inline lib/Site.js]: This should be above `FOOTER_PATH` for alphabetical order.

File: lib/Site.js [hunk 2]
```diff
@@ -132,16 +133,25 @@ function rejectHandler(reject, error, removeFolders) {
...
+    const faName = fontAwesomeClass.replace(' fa', '').replace(/-/g, '_');
+    const faHtml = `<span class='${fontAwesomeClass}'></span>`;
+    iconsMap[faName] = faHtml;
+  });
+  // Collate Glyphicons
+  glyphiconsClasses.forEach((glyphiconClass) => {
+    const glyphName = glyphiconClass.replace(/-/g, '_');
+    const glyphHtml = `<span class='glyphicon ${glyphiconClass}' aria-hidden='true'></span>`;
+    iconsMap[glyphName] = glyphHtml;
+  });
```
Comment (yamgent) [inline lib/Site.js]: Was wondering whether there's any value in refactoring the code since it seems the logic for "Font Awesome" and "Glyphicons" are almost identical.

File: docs/userGuide/components/panel.md [hunk 3]
```diff
@@ -64,13 +64,32 @@
 
 <tip-box border-left-color="#00B0F0">
   <i style="font-style: normal; font-weight: bold; color: dimgray">Example</i><br>
-  <code>primary</code> type<br>
-  <panel header="#### primary type panel" type="primary">
+  <p>Click the Panels to see the expanded style.<p>
```
Comment (yamgent) [inline docs/userGuide/components/panel.md]: Er... the commit `Docs: Update panel section` doesn't seem to be related to this PR? :P

File: docs/userGuide/contentAuthoring.md [hunk 4]
```diff
@@ -125,6 +125,29 @@ You can insert variables to be replaced by pre-defined values anywhere in your s
...
+Note that you will have to shorten the icon's name before using it within MarkBind. It is a two step process: 
+1. Shift the icon's prefix to replace `fa`, the icon's name first two characters. e.g `fas fa-chess-knight` becomes `fas-chess-knight`
+2. Next, replace all instances of `-` with `_` in the icon's name. e.g. `fas-chess-knight` becomes `fas_chess_knight`
+
+After doing so, you are able to use the icons anywhere on your site by surrounding an icon's shortened name with double curly braces.<br>
+e.g. <code>{<span></span>{fas_fighter_jet}}</code> will be rendered as {{fas_fighter_jet}}
+
 #### Glyphicons
 
 This asset is provided by [Glyphicons](https://glyphicons.com/) via Bootstrap.
```
Comment (yamgent) [inline docs/userGuide/contentAuthoring.md]: Let's specify the version number of Bootstrap here since there's now multiple versions of Bootstrap floating around.

"Bootstrap" -> "Bootstrap 3".

File: docs/userGuide/contentAuthoring.md [hunk 5]
```diff
@@ -125,6 +125,29 @@ You can insert variables to be replaced by pre-defined values anywhere in your s
 
 Some helpful variables are provided by MarkBind for your convenience, as shown below. 
 
+#### Font Awesome Icons
+
+This asset is provided by [Font Awesome](https://fontawesome.com/) under their [free license](https://fontawesome.com/license).
+<br><br>
```
Comment (yamgent) [inline docs/userGuide/contentAuthoring.md]: I see that without the `<br><br>`, the code below is rendered literally.

I was able to avoid using `<br><br>` by specifying the code below as `<markdown>` explicitly. Not sure which one is a better option though:

```html
<tip-box>
Traditional usage:  
<markdown>`<span class='fas fa-chess-knight'></span>`</markdown>
<br>  
MarkBind usage:
<br>
<code>{<span></span>{fas_chess_knight}}</code>
</tip-box>
```

File: docs/userGuide/contentAuthoring.md [hunk 6]
```diff
@@ -125,6 +125,29 @@ You can insert variables to be replaced by pre-defined values anywhere in your s
...
+Traditional usage:  
+`<span class='fas fa-chess-knight'></span>`
+<br>  
+MarkBind usage:  
+<code>{<span></span>{fas_chess_knight}}</code>
+</tip-box>
+
+Font Awesome has three different styles for their free icons, each with their own prefix:
+<strong>Solid</strong> (`fas`), <strong>Regular</strong> (`far`) and <strong>Brands</strong> (`fab`).<br>
+Be sure to check Font Awesome's [full list of icons](https://fontawesome.com/icons?d=gallery&m=free) on which prefix to use.<br>
```
Comment (yamgent) [inline docs/userGuide/contentAuthoring.md]: "on which prefix to use." Maybe should just say for "the full name of the icon" or similar?

Also it can be easily confused by what we meant by prefix --> "fas " and "fa-" are both techincally prefixes.

File: docs/userGuide/contentAuthoring.md [hunk 7]
```diff
@@ -137,10 +160,13 @@ MarkBind usage:
 <code>{<span></span>{glyphicon_hand_right}}</code>
 </tip-box>
 
-Use these [available glyphs](https://getbootstrap.com/docs/3.3/components/#glyphicons) on your site by surrounding a glyph's name with double curly braces.  
-e.g. <code>{<span></span>{glyphicon_hand_right}}</code> will be rendered as <code><span class='glyphicon glyphicon-hand-right' aria-hidden='true'></span></code>
+Here is the [list of provided glyphs](https://getbootstrap.com/docs/3.3/components/#glyphicons) from Bootstrap.
+
+Note that you will have to modify the glyph's name first before being able to use it within MarkBind.<br>It is done by replacing all instances of `-` with `_` in the glyph's name.
```
Comment (yamgent) [inline docs/userGuide/contentAuthoring.md]: The next line is an example, but this statement does not have a `<br>` at the end, unlike the others.

File: docs/userGuide/components/panel.md (hunk 3 above)
Comment (Chng-Zhi-Xuan) [reply docs/userGuide/components/panel.md]: As there were new types introduced, this section has to be updated. I did some extra formatting to make it more explicit in presenting the different styles.

File: lib/Site.js (hunk 2 above)
Comment (Chng-Zhi-Xuan) [reply lib/Site.js]: It is difficult to abstract into a single function as they have different algorithms for extracting their variable name and the html formatting is also different.

File: docs/userGuide/contentAuthoring.md (hunk 6 above)
Comment (Chng-Zhi-Xuan) [reply docs/userGuide/contentAuthoring.md]: The message I was trying to convey was that some icons do not have certain styles.

Maybe change to `"... full list of icons and which styles they support."`

File: docs/userGuide/contentAuthoring.md (hunk 5 above)
Comment (Chng-Zhi-Xuan) [reply docs/userGuide/contentAuthoring.md]: The `<br>` tag between the 2 usage blocks is not needed. Looks cleaner when using the `<markdown>` tag.

**Comment by yamgent** [comment]
> Acknowledgements of any other projects or assets we use will be done in our repository's README page.

Will you be doing that for Bootstrap 3 in this PR?

**Comment by Chng-Zhi-Xuan** [comment]
> Will you be doing that for Bootstrap 3 in this PR?

All acknowledgements including Bootstrap 3 should be in a separate PR and issue. Not directly related in this PR.

File: docs/userGuide/contentAuthoring.md [hunk 8]
```diff
@@ -125,22 +125,48 @@ You can insert variables to be replaced by pre-defined values anywhere in your s
...
+MarkBind usage:
+<br>
+<code>{<span></span>{fas_chess_knight}}</code>
+</tip-box>
+
+Font Awesome has three different styles for their free icons, each with their own prefix:
+<strong>Solid</strong> (`fas`), <strong>Regular</strong> (`far`) and <strong>Brands</strong> (`fab`).<br>
+Be sure to check Font Awesome's [full list of icons](https://fontawesome.com/icons?d=gallery&m=free) and which styles each icon supports.
+
+Note that you will have to shorten the icon's name before using it within MarkBind. It is a two step process:
```
Comment (acjh) [inline docs/userGuide/contentAuthoring.md]: - "you will *have to* ... before using it" sounds like a chore — and authors don't *have to* use MarkBind's way.
  How about: "MarkBind offers a shorter syntax by transforming the icon's name. It is a two-step process:"

- Use similar wording for Glyphicons.

File: lib/Site.js [hunk 9]
```diff
@@ -132,16 +133,25 @@ function rejectHandler(reject, error, removeFolders) {
...
   const glyphiconsClasses = fs.readFileSync(glyphiconsPath, 'utf8').trimRight().split(/\r?\n/);
-  return glyphiconsClasses.reduce((glyphiconsMap, glyphiconClass) => {
-    const name = glyphiconClass.replace(/-/g, '_');
-    const html = `<span class='glyphicon ${glyphiconClass}' aria-hidden='true'></span>`;
-    // eslint-disable-next-line no-param-reassign
-    glyphiconsMap[name] = html;
-    return glyphiconsMap;
-  }, {});
+  const iconsMap = {};
+  // Collate Font Awesome Icons
```
Comment (acjh) [inline lib/Site.js]: Unnecessary comments (same for Glyphicons).

File: lib/Site.js [hunk 10]
```diff
@@ -132,16 +133,25 @@ function rejectHandler(reject, error, removeFolders) {
...
+  fontAwesomeClasses.forEach((fontAwesomeClass) => {
+    const faName = fontAwesomeClass.replace(' fa', '').replace(/-/g, '_');
+    const faHtml = `<span class='${fontAwesomeClass}'></span>`;
+    iconsMap[faName] = faHtml;
```
Comment (acjh) [inline lib/Site.js]: No need prefix:

```js
const name = ...
const html = ...
iconsMap[name] = html;
```

File: lib/Site.js [hunk 11]
```diff
@@ -264,6 +274,10 @@ Site.prototype.createPage = function (config) {
                                path.join(this.siteAssetsDestPath, 'css', 'bootstrap.min.css')),
       bootstrapVue: path.relative(path.dirname(resultPath),
                                   path.join(this.siteAssetsDestPath, 'css', 'bootstrap-vue.min.css')),
+      fontAwesome: path.relative(path.dirname(resultPath),
+                                 path.join(this.siteAssetsDestPath, 'css', 'font-awesome.css')),
+      glyphicons: path.relative(path.dirname(resultPath),
+                                path.join(this.siteAssetsDestPath, 'css', 'glyphicons.css')),
```
Comment (acjh) [inline lib/Site.js]: - Let's provide and use the minified versions.
- Let's name it **bootstrap-glyphicons.css** (and **.min.css**) to be explicit about its source.
